crates/hir-ty/src/consteval.rs

```diff
@@ -269,13 +277,17 @@ pub(crate) fn const_eval_discriminant_variant(
         };
         return Ok(value);
     }
+
+    let repr = db.enum_data(loc.parent).repr;
+    let is_signed = repr.and_then(|repr| repr.int).is_none_or(|int| int.is_signed());
+
     let mir_body = db.monomorphized_mir_body(
         def,
         Substitution::empty(Interner),
         db.trait_environment_for_body(def),
     )?;
     let c = interpret_mir(db, mir_body, false, None).0?;
-    let c = try_const_usize(db, &c).unwrap() as i128;
+    let c = try_const_usize_sign_extend(db, &c, is_signed).unwrap() as i128;
```

Comment:
Previously, the discriminant was always computed as being positive. This affects this assist and the enum discriminant inlay hint.